sklearn/neighbors/base.py

```diff
@@ -99,6 +98,55 @@ def _get_weights(dist, weights):
                          "'distance', or a callable function")
 
 
+def _decompose_neighbors_graph(graph, accept_unequal=True):
+    """Decompose a nearest neighbors sparse graph into distances and indices
+
+    Parameters
+    ----------
+    graph : CSR sparse matrix
+        Neighbors graph as given by kneighbors_graph or radius_neighbors_graph
+
+    accept_unequal : boolean, default=True
+        Whether to accept graphs with different number of neighbors for each
+        sample.
+
+    Returns
+    -------
+    distances : array, shape (n_samples, n_neighbors) or array of arrays
+        Distances to nearest neighbors.
+        It is an array of arrays if the number of neighbors is not identical
+        for every samples.
+
+    indices : array, shape (n_samples, n_neighbors) or array of arrays
+        Indices of nearest neighbors.
+        It is an array of arrays if the number of neighbors is not identical
+        for every samples.
+    """
+    n_samples = graph.shape[0]
+
+    # if there is the same number of neighbors for each sample
+    if np.unique(np.diff(graph.indptr)).size == 1:
+        distances = graph.data.reshape(n_samples, -1)
+        indices = graph.indices.reshape(n_samples, -1)
+        return distances, indices
+
+    else:  # radius neighbors case
+        if not accept_unequal:
+            raise ValueError(
+                'Precomputed sparse neighbors graph does not have the same '
```

Comment:
I don't think this should be a requirement.

Reply:
The user should be welcome to overgenerate nearest neighbors then use it for getting multiple subsets of different sizes (or bound by radius rather than number).

Reply:
I agree the user should be able to generate `k` nearest neighbors and then use only the `k'` nearest (with `k'<k`), then `k''` nearest, etc. Same thing with the radius query.

But would you say the user should also be able to generate `k` neighbors and then use it with a radius query `r`, and the opposite, generating the neighbors closer than `r` and then use the `k` nearest.

Reply:
Yes, I think so. Or at least I don't see why not, and stopping them from doing so would be harder than letting them...